Add panel-dsi for device-tree side configuration of generic DSI panels

timonsku opened the pull request:
This adds an analog to the already existing panel-dpi to panel-simple which allows for timing configuration in device-tree for simple DSI panels.
In addition this also allows configuration of the DSI parameters in device tree.
I tried to be as consistent as possible with how panel-dpi was implemented.
Code was tested with a 840x480 panel driven by an externally configured bridge IC.

This addition was discussed with @6by9 over in https://github.com/raspberrypi/firmware/issues/1819#issuecomment-1725769614
The main motivation is with not needing to run custom kernels for simple DSI devices that require nothing but an entry in panel-simple for the timings and DSI config flags.
This becomes very cumbersome when a large amount of displays need to be maintained. Essentially the exact same reasoning that the panel-dpi patch had for the initial implementation for this type of feature in panel-simple but expanding it to also be possible for DSI panels.

Any feedback would be welcome.

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -4681,23 +4684,123 @@ static const struct of_device_id dsi_of_match[] = {
+	}
+
+	of_property_read_string(np, "mode", &dsi_mode_flags);
+	if (!strcmp(dsi_mode_flags, "MODE_VIDEO"))
```
The MIPI_DSI_MODE_xxx flags are generally used in combination, not mutually exclusive.
eg MODE_VIDEO switches between command mode and video mode displays. Almost all the others are only applicable to video mode, so have to have that set as well.

If your example binding with
```
mode = "MODE_VIDEO", "MODE_VIDEO_BURST", "MODE_NO_EOT_PACKET";
```
is valid, then that is great. However I don't see this parsing code reading all of those flags.

pelwell replied:
Hint: Look at `of_property_for_each_strings`, which uses the length of the property to decide if there may be more strings.

Aside: I find the following related code in `of_prop_next_string` worrying:
```
	curv += strlen(cur) + 1;
	if (curv >= prop->value + prop->length)
		return NULL;
```
The second line acknowledges that the last string may not be NUL-terminated by comparing the apparent length of the string with the overall length of the property, but by then it may be too late - the strlen may already have stumbled into an invalid page. I should work up the enthusiasm to raise the matter upstream.

timonsku replied:
Ooh wow yea that fell of my todo, thanks for spotting that. My panel for testing only requires MODE_VIDEO hence working fine despite this bug..

timonsku replied:
Added in last commit.

6by9 commented on Documentation/devicetree/bindings/display/panel/panel-dsi.yaml:
```diff
@@ -0,0 +1,111 @@
+    items:
+      - {}
+      - const: panel-dsi
+
```
Referencing https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/display/panel/panel-simple-dsi.yaml#L58-L60, you ought to be supporting a reg property to set the DSI virtual channel.

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -4681,23 +4684,123 @@ static const struct of_device_id dsi_of_match[] = {
-	desc = id->data;
+	if (id->data == &panel_dsi) {
+		/* Handle the generic panel-dsi binding */
+		err = panel_dsi_dt_probe(&dsi->dev, dt_desc);
```
The handlng of dt_desc here looks suspicious.

Defined as an uninitialised pointer, that uninitialised value is passed to panel_dsi_dt_probe which dereferences it and writes the desc field. We then store that uninitialised value in desc for use later :-/

timonsku replied:
I might be missing what you mean but panel_dsi_dt_probe allocates the desc struct that gets assigned?

timonsku replied:
Oh wait yea no I think you have a point, the outer member pointer itself does not get allocated. This might be running on undefined behaviour right now, otherwise it should blow up which it weirdly doesn't atm.

What would you think about making dt_desc static? My main gripe there was that the panel definitions are supposed to be treated as const there so I can't reuse panel_dsi which would be the better way really.

6by9 replied:
You can't reuse `panel_dsi` as you could have 2 displays configured through the same `panel-dsi` compatible.

You are right that there is a conflict that `panel_simple_probe` gets a dummy desc as panel_dpi, and then `panel_dpi_probe` handles allocating the desc. That bit almost does want to be pulled out in order to make it potentially common.

Your current implementation does all the work before `panel_simple_probe` gets a look in.

timonsku replied:
Oh right, I guess that might need fixing then too for panel-dpi, that is currently re-useing the single dummy struct :)
Though I'm not sure how you would handle multiple panels like this properly that does not rely on hard coded dummy panels to refer to. I think that would require some deeper changes to remove the assumption that a panel is in some way already statically defined. I might need to test this if it might already work unintentionally, the whole driver is a bit of a rabbit hole of multiple layers of structs being passed around and torn apart.

timonsku replied:
Got some DSI cables for PI5 today so I could test multiple panels working correctly.
That works well with the latest commit.
Regarding my earlier comment in the other thread. While I think that is probably the best approach down the line I found that it becomes quite far reaching to add a general "get timing from dt" functionality in that isn't port specific like right now.
I think that would be something that should probably be tested across more devices and I don't currently have hardware here to also test this with LVDS.
Keeping the ~15 duplicate lines feels like the better compromise for now as its otherwise awkward to merge just the dpi and dsi specific probing bits.
Let me know what you think of this.

6by9 replied:
panel_dpi is fine as `desc` gets swapped for `panel->desc`, which is allocated in `panel_dpi_probe`.

Now that you devm_kzalloc the desc, I think you're OK too.

6by9 replied:
I would like to see the duplication removed, but only if you're committing to upstream the differences.
Handling major changes to files such as this that get a fair amount of churn is not easy.

timonsku replied:
As in for an upstream submission you would like to see the larger change that abstracts the device tree side timing definition?

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -4681,23 +4684,123 @@ static const struct of_device_id dsi_of_match[] = {
+	desc->num_timings = 1;
+
+	of_property_read_u32(np, "width-mm", &desc->size.width);
+	of_property_read_u32(np, "height-mm", &desc->size.height);
```
Up to here in this function is all common with panel_dpi_probe.

If we change the 
```
if (desc == &panel_dpi) {
```
at https://github.com/torvalds/linux/blob/master/drivers/gpu/drm/panel/panel-simple.c#L593 to
```
if (desc == &panel_dpi || desc == &panel_dsi) {
```
don't you get that for free?

timonsku replied:
The flow of DSI panels is quite seperate to all the other types. What you propose is what I had done previously but then decided against it as it breaks with how the driver is currently constructed (note that two drivers are actually registered, one for panel-simple and one for panel-simple-dsi).
I'm unsure tbh what the best approach is but yes there is a fair chunk of duplicate code but panel_dpi_probe assumes a dpi panel and writes dpi specific properties.
You could to some degree merge these two probe functions but it would require then two seperate spaces where we parse the device tree, one for the DSI port specific aspects set in panel_desc_dsi in panel_simple_dsi_probe and in panel_dpi probe with some if else required to set the dpi or dsi specific settings of panel_desc.

Separating the two despite the overlap in function and following the currently set flow of panel-simple-dsi seemed like the better approach.
Generally speaking panel-simple might require a broader restructuring as there are two competing ideas in there.
If I were to do that I would remove all of the existing panel definitions, move them into device tree overlays and have the driver be fully device tree defined but I don't think that would be acceptable as it would be a breaking change and given how contentious this was in the past would probably mean a lengthy discussion with upstream first.

6by9 replied:
`panel_dpi_probe` currently takes `*dev` and a `struct panel_simple *panel`. It only uses `*panel` to fill in the `desc` field.

Modify it to take a `struct panel_desc **desc`, and pass in &panel-desc from both dpi and dsi paths. The DSI probe can then continue and read the DSI specific bits.

Yes it would leave a couple of minor and inconsequential DPI related things executed in the DSI path, but if viewed necessary those could be split out.

timonsku replied:
Ok! So my approach so far was to only add and not change anything. The only reason for that was to appease to an upstreaming process as much as possible. If you think it would be fine to make some changes to the existing things then I would propose to keep the probing functions seperate but pull out the panel timing and panel-common property parsing into a new function that takes on the same role as the EDID parsing does. That function adds additional timings to a panel that were found on the EDID connected to the DT defined I2C port.
That feels like a more generic way and would allow any type of "connector" to get timings from device tree and the same wouldn't need to be done for say LVDS.
I'm a bit unsure about the dynamic creation of a panel as discussed above though. Could stick with having some dummy structs for now or see to handle this a bit more properly. A bit of a question of scope I guess, there is a lot that could technically be improved overall.
What are your thoughts?

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -500,6 +500,7 @@ static int panel_dpi_probe(struct device *dev,
 	return 0;
 }
 
+
```
Random whitespace change

6by9 replied:
Still present

aBUGSworstnightmare-rpi commented:
May I ask which bridge device this is needed/intended for?
- for a MIPI-to-DPI bridge (as used on the official 7in) a "panel-dpi" compatible is there already 
		panel_generic: __overlay__  {
			compatible = "panel-dpi";

			width-mm = <154>;
			height-mm = <83>;
			bus-format = <0x1009>;

			timing: panel-timing {
				clock-frequency = <29500000>;
				hactive = <800>;
				hfront-porch = <24>;
				hsync-len = <72>;
				hback-porch = <96>;
				hsync-active = <1>;
				vactive = <480>;
				vfront-porch = <3>;
				vsync-len = <10>;
				vback-porch = <7>;
				vsync-active = <1>;

				de-active = <1>;
				pixelclk-active = <1>;
			};
		};

- for a MIPI-to-LVDS bridge one needs to enable 'Generic LVDS panel driver' in order to use  "panel-lvds" compatible from DT
- i.e.
- `			panel: panel {
				compatible = "panel-lvds";

				backlight = <&backlight_lvds>;

				/* Physical dimensions of active area */
				width-mm = <217>;
				height-mm = <136>;

				data-mapping = "vesa-24";

				panel-timing {
					clock-frequency = <71100000>;
					hactive = <1280>;
					hfront-porch = <160>;	
					hsync-len = <48>;
					hback-porch = <48>;
					
					vactive = <800>;
					vfront-porch = <23>;
					vsync-len = <6>;					
					vback-porch = <8>;
				};`

Bridge input timing will have to be calculated by the driver based on the required output (in most cases) and other parameters (i.e. no data-lanes can be configured via overrides).

so, what is the real target here? Having the possibility for adding 'dumb panels' (panels which just require a suitable timing on the DSI interface)? But that would not be in-line with

> https://github.com/raspberrypi/firmware/issues/1819
I'm currently working on product for Adafruit that would support different panels via a bridge IC like it is used in the official touchscreen.

Would be glad if you can share some more details so others might be able to support.

timonsku commented:
@aBUGSworstnightmare-rpi 
There are multiple things where this would be needed. The use case mentioned in the firmware issue is one of them but does not really relate to the IC itself, that could be any bridge really but yes its a DSI to DPI bridge IC in that instance but might also change down the line which one specifically or new products with different bridges being added where its desirable to keep software compatibility so that the process for the end user looks the same, so the bootstrapping would happen externally and the requirement to the host would only be to send a suitable DSI stream.
A fixed port configuration would not be a great solution as different applications have different needs in terms of DSI bus flags.

"Native" panels without a bridge IC are also a target. The wish is really only to be able to define what one can define in a panel_desc_dsi but doing that at runtime without needing to get a patch merged or distribute a kernel fork.
There are potentially dozens or even hundreds of different panels in the future and defining them all in the driver is just not sustainable, again exact same motivation as with the original pane-dpi patch that enabled DT timings but wanting control over the DSI bits too.

aBUGSworstnightmare-rpi commented:
> "Native" panels without a bridge IC are also a target.

Hmm .. still don't get that!
A 'native' DSI for me is a command mode display. Find a typical power-on-sequence example from the image linked here [https://drive.google.com/file/d/1Vi3UZlcxU89LIjbEj2nI8zCIRfpN5BTa/view?usp=sharing](url)

You can't fire this display up with a panel-simple-dsi compatible; you will have to use panel-dpi-cm. That's pretty much the same for all native DSI displays which autoload TCON register settings from NVM; one still has to send 0x51/0x53/0x55 before you can switch the display on (0x29) and exit sleep mode (0x11) for normal operation. 
Looking at the sequence, a 'dumb' display simply needs 'Host Display Data transfer'

6by9 commented:
Native DSI can be either command mode or video mode.
I've never encountered a command mode panel in reality and have no idea how well (or not) they'll behave. When someone actually finds one, I'll worry about it.

Whilst I wouldn't advocate it, the Waveshare DSI panels could be considered native DSI panels because the MCU handles all the initialisation. The regulator and backlight side could be handled by a regulator driver similar to rpi_panel_attiny_regulator, which leaves the panel side as a simple DSI device. That is a valid use case in my book.

I would advocate that bridge chips are configured from Linux rather than hidden behind an MCU. At least that way there is visibility over what the hardware is actually doing, and if a compatibility issue is observed with a future chip then it can be addressed. Most bridge chip drivers should be configured automatically by the upstream panel device, so other than a bit of DT it adds no overhead.

timonsku commented:
@aBUGSworstnightmare-rpi see the panels in the `dsi_of_match` array for examples of DSI video mode panels that do no require initialization.
And what 6by9 described, sometimes its useful to init externally for either systems reasons or because the init sequence is inexplicably under NDA. In those cases you can turn them into simple panels.

@6by9 generally agree with the sentiment about bridges! It gets a bit more complicated in some specific dev board style applications as I found out as the needs break with some assumptions in existing drivers or guidelines of how drivers should behave. That is a very niche thing though where I feel re-using generic infrastructure might be the nicer approach.

ladyada commented:
(subscribing myself by commenting)

pelwell commented:
That works - there's also a Subscribe button in the controls on the right at the top of the page (it probably says Unsubscribe now).

pelwell commented:
Can you stop force-pushing your branch unless you are actively changing your code? GitHub is pretty good at rebasing when we want to merge, and every update invalidates the previous build check results.

timonsku commented:
sorry, there were commits that were not supposed to go into this branch that I wanted to roll back. Not going to happen again.

arturo182 commented:
Just wanted to add that I tested this PR on the CM4 with a custom carrier I'm working on, and it works great! I no longer have to patch `panel-simple.c` to add my custom panel :)

ladyada commented:
@pelwell ive also had much success with fun displays connecting through the ICN6211 (with a microcontroller doing the i2c/spi configuration during power-up) - can you approve the workflow so it could get reviewed before the next kernel release is made? :)

![image](https://github.com/raspberrypi/linux/assets/1214161/d6be51f5-83fb-4b5c-964c-eadce09b52db)

pelwell commented:
Is this actually ready? I see that the Draft status has been removed (GitHub doesn't give notifications about that, so a comment is always a good idea), but some of the comments on the PR have been acknowledged but not acted on.

6by9 commented on Documentation/devicetree/bindings/display/panel/panel-dsi.yaml:
```diff
@@ -0,0 +1,114 @@
+  - |
+    panel {
+        compatible = "panel-mfgr,generic-dsi-panel","panel-dsi";
+        label = "Simple DSI Panel";
```
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/display/panel/panel-common.yaml#L34-L46
>       If no such name exists, and unless the property is mandatory according to
>      device tree bindings, it shall rather be omitted than constructed of
>      non-descriptive information. For instance an LCD panel in a system that
>      contains a single panel shall not be labelled "LCD" if that name is not
>      inscribed on the system or used in a descriptive fashion in system
>      documentation.

timonsku replied:
right what I was trying to convey is that this should be replaced with actual panel information by the user as its done with the dpi binding. Though for the `Label` property thats probably not obvious enough.
I think it is useful to have them so that they show up as the actual panel name but of course should be filled out properly. So f.e.
```
        compatible = "TDO,TL040HD","panel-dsi";
        label = "TL040HD";
```

6by9 replied:
AFAIK it's never passed on to the user, and it's more for those analysing the DT to be able to understand which device is being derived.
Looking at the other bindings, very few define a label (amusingly panel/lgphilips,lb035q02.yaml includes `label = "lcd";` almost to spite the writer of the docs!), and a very quick grep in arch/arm/boot/dts doesn't appear to include any displays with labels.

Please drop it.

timonsku replied:
thats fair, was mostly following the other panel-*.yaml bindings which make use of it but you are right its actually not passed to user space. I somehow remembered that it would be used for the display name in the window manager but that does not seem to be the case, at least not with Wayland. Does not have any value in that case.

6by9 commented on Documentation/devicetree/bindings/display/panel/panel-dsi.yaml:
```diff
@@ -0,0 +1,114 @@
+            hsync-active = <0>;
+            vsync-active = <0>;
+            de-active = <1>;
+            pixelclk-active = <1>;
```
None of these 4 are applicable to DSI so should be omitted.

6by9 commented on Documentation/devicetree/bindings/display/panel/panel-dsi.yaml:
```diff
@@ -0,0 +1,114 @@
+    description: |
+      The number of lanes used by the panel.
+    $ref: /schemas/types.yaml#/definitions/uint32
+    enum: [1, 2, 3, 4]
```
The more consistent approach would be to use data-lanes as an array of data lane indexes. This is used across a number of bridges, and is consistent with V4L2.
eg https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/display/bridge/chipone,icn6211.yaml#L64-L71

timonsku replied:
I don't quite understand the use of that format. The intended use seems to then be `drm_of_get_data_lanes_count_ep` to get the number of elements in the array to get the number that you currently just set directly. That seems very convoluted?
Happy to change it if its the canonical way to set DSI lane count, just feels quite odd but maybe I'm missing an aspect here. If there are PHYs that support picking specific lanes the dsi_panel struct does not support that at least as its just an unsigned int. Not sure if its good to convey a support for that if the information is just dropped.

6by9 replied:
I'll admit that DRM is less consistent than V4L2, but some devices support reordering of the data lanes. Using the array allows for a consistent way of configuring the interface.

Any non-standard property is meant to be prefixed with a vendor code. `data-lanes` is standardised. `lanes` is not. Friendly advice if upstreaming - use the standard properties.

timonsku replied:
Thats fair yea. I will change it to `data-lanes`.

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -4681,31 +4685,143 @@ static const struct of_device_id dsi_of_match[] = {
+		err = panel_dsi_dt_probe(&dsi->dev, dt_desc);
+		if (err < 0)
+			return err;
+		err = panel_simple_probe(&dsi->dev, &dt_desc->desc);
```
Massive duplication here.
```
if (id->data == &panel_dsi) {
  ...
   desc = &dt_desc->desc;
} else {
   desc = id->data;
}

err = panel_simple_probe(...);
...
```

timonsku replied:
Yes, having a second a think on this I realize the reason I did this actually has a different solution that wouldn't require this. Will try to remove it.

6by9 reviewed (changes requested):
Sorry, still unaddressed issues.

There needs to be 2 commits - one device tree binding, and one driver change. The third "fixup" patch touches both binding and driver.

Checkpatch complaint as there is no commit description.

How about a 3rd commit that adds an overlay using this, along similar lines to https://github.com/raspberrypi/linux/blob/rpi-6.1.y/arch/arm/boot/dts/overlays/vc4-kms-dpi-generic-overlay.dts or vc4-kms-dpi-panel-overlay.dts

For those adding panels using this driver change, I would like to see it done in a similar way to https://github.com/raspberrypi/linux/blob/rpi-6.1.y/arch/arm/boot/dts/overlays/vc4-kms-dpi-panel-overlay.dts, where there is one overlay that uses an override to select the relevant panel. The timing and config can be put in that overlay.

timonsku commented:
Thanks for the input @6by9 !
Regarding splitting binding and driver patch. I thought the initial commit of a new binding should be split into its own commit (which I did).
So what you are saying I should always keep changes to bindings completely separated too? Should I then revert the latest (third) commit and split it into two (making a total of 4 excluding upcoming fixes for the other things you addressed)

For the requested overlay, you mean this a practical example for how to use this?

6by9 commented:
> Thanks for the input @6by9 ! Regarding splitting binding and driver patch. I thought the initial commit of a new binding should be split into its own commit (which I did). So what you are saying I should always keep changes to bindings completely separated too? Should I then revert the latest (third) commit and split it into two (making a total of 4 excluding upcoming fixes for the other things you addressed)

With a very few exceptions, individual commits should never cross subsystem boundaries. For mainline it causes grief over which tree it should be committed through. 

We also don't want your development history here - you've been asked to make changes as the original driver has issues, therefore please amend the existing commits as they've gone nowhere except your local branch.

> For the requested overlay, you mean this as a practical example for how to use this?

Yes please.

timonsku commented:
> Is this actually ready? I see that the Draft status has been removed (GitHub doesn't give notifications about that, so a comment is always a good idea), but some of the comments on the PR have been acknowledged but not acted on.

Sorry I just realized that my added comments where part of making a review and did not show up publicly as I hadn't submitted the review until now.

Commit:
overlays: example overlay for using panel-dsi on RPi

Analog to the generic panel-dpi overlay to use panel-dsi with dtparam
to not require a panel specific overlay for simple use cases that
do not require setting more niche DSI modes.

Signed-off-by: Timon Skerutsch <kernel@diodes-delight.com>

Commit:
drm/panel: add panel-dsi

Equivalent to panel-dpi for configuring a simple DSI panel with
device tree side timings and bus settings.
Motiviation is the same as for panel-dpi of wanting to support
new simple panels without needing to patch the kernel.

Signed-off-by: Timon Skerutsch <kernel@diodes-delight.com>

Commit:
dt-bindings: display: panel-dsi bindings

Bindings for the panel-dsi specific additions to panel-simple.
Allow for DSI specific bus settings and panel timing
to be define in devicetree. Very similar to panel-dpi.

Signed-off-by: Timon Skerutsch <kernel@diodes-delight.com>

timonsku commented:
Sorry for the force push but not aware of any other way to make a clean history rewrite/amend.
I think I covered all the outstanding change requests. A third commit added for an example overlay.

6by9 commented on Documentation/devicetree/bindings/display/panel/panel-dsi.yaml:
```diff
@@ -0,0 +1,111 @@
+        backlight = <&backlight>;
+        dsi-color-format = "RGB888";
+        reg = <0>;
+        data-lanes = <1 2>;
```
This should be under the endpoint, not the top level of the panel.

Definition should also match.

timonsku replied:
I guess relates to the other comment. Would you then put all the dsi properties under the endpoint?
Feels inconsistent otherwise to me if the dsi specific bits are set in different places.
Though then not really sure how that would be defined in a binding when it essentially touches the binding of something it only inherits?

6by9 commented on arch/arm/boot/dts/overlays/panel-simple-dsi.dts:
```diff
@@ -0,0 +1,55 @@
+                    };
+            };
+            panel@0 {
+                // See panel-dsi.yml binding for more information
```
.yaml

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -4681,23 +4685,141 @@ static const struct of_device_id dsi_of_match[] = {
+	of_property_read_u32(np, "width-mm", &desc->size.width);
+	of_property_read_u32(np, "height-mm", &desc->size.height);
+
+	dsi_lanes = drm_of_get_data_lanes_count(np, 1, 4);
```
Prototype of drm_of_get_data_lanes_count is 
```
int drm_of_get_data_lanes_count(const struct device_node *endpoint,
				const unsigned int min, const unsigned int max)
```
You've passed the device_node of the base device, not the endpoint.

Use ```drm_of_get_data_lanes_count_ep(np, 0, -1, 1, 4);```

timonsku replied:
the property does not sit in the endpoint though but is part of the panel-dsi binding to be able to pass that into the dsi_mode for setup.
`drm_of_get_data_lanes_count_ep` would not be able to find the property then. If you check the implementation of the functions the use of `endpoint` of the non `_ep` prototype for it is a bit misleading because it does not actually expect an endpoint it just looks for the data-lanes property within the node it was given and does a boundary check, nothing specific to endpoints in there.
`drm_of_get_data_lanes_count_ep` does look for an endpoint though and expects to find the property there.
So I think in this case `drm_of_get_data_lanes_count` is more apropriate. Otherwise I would not know how to specify the binding properly without modifying the inherited bindings.

timonsku replied:
switched data-lanes to be in endpoint

6by9 commented on drivers/gpu/drm/panel/panel-simple.c:
```diff
@@ -4681,23 +4685,141 @@ static const struct of_device_id dsi_of_match[] = {
+		desc_dsi->lanes = dsi_lanes;
+	}
+
+	of_property_read_u32(np, "lanes", &desc_dsi->lanes);
```
Why are we still trying to read this?

timonsku replied:
oh yea thats not supposed to still be there, may have clicked the wrong thing during rebase

6by9 reviewed (commented):
Force pushing is the correct way to update the branch backing a PR.

Thank you - the general implementation is far cleaner.

A number of checkpatch errors.

Please give a full description in the commit text. Titles should always start with the subsystem (do `git log` on the file if in doubt, and copy all the other commits).

I was hoping that the overlay would actually be compiled and useful, with similar overrides to https://github.com/raspberrypi/linux/blob/rpi-6.1.y/arch/arm/boot/dts/overlays/vc4-kms-dpi-generic-overlay.dts

timonsku commented:
Ah yes sorry about the missing description you mentioned that before, will run checkpatch on the actual patch. I only ran it on the source files.

@ Overlay, my bad I understood that way different. More as a template for new overlays, not to be used as is.
But yes that makes sense, so the intention is full configuration with overrides so that panels don't even necessarily have to have an overlay installed? (e.g. same function as the dpi overlay)

timonsku commented:
@6by9 do you have a suggestion for how to represent the mode stringarray with overrides?
As I understand there is currently no direct support for array types in the overrides syntax and I'm not aware of any devicetree "native" way of appending values to an array with overlays (in which case I could use fragments).
The only way I currently see is having a fragment for every possible combination but that is not practical.

pelwell commented:
We can do arrays of cells, but not arrays of strings. Exactly what change do you want the overlay to make?

6by9 commented:
> @6by9 do you have a suggestion for how to represent the mode stringarray with overrides? As I understand there is currently no direct support for array types in the overrides syntax and I'm not aware of any devicetree "native" way of appending values to an array with overlays (in which case I could use fragments). The only way I currently see is having a fragment for every possible combination but that is not practical.

- Seeing as we don't really support command mode (I've never seen a command mode panel in the flesh), you can assume MODE_VIDEO is set. Command mode panels would be supported through [panel-dsi-cm](https://github.com/raspberrypi/linux/blob/rpi-6.1.y/drivers/gpu/drm/panel/panel-dsi-cm.c) anyway, and not panel-simple.
- MODE_LPM is redundant as this is the simple panel-dsi that sends no commands.
- HS_PKT_END_ALIGNED is a Mediatek special.
- And actually the vc4 driver currently doesn't look at the other flags, so all of them can be ignored.

So just set MODE_VIDEO and have done with it.

If we add support for the extra flags, then a vc4-kms-dsi-panel-overlay.dts overlay would allow those who wish to define their panel by name can do support that level of configuration, but the generic overlay won't support it. Overlays don't go upstream, so it's not an issue for that.

timonsku commented:
That's fair enough, good with me. I don't see a big need for more than video_mode either. Most other options are usually just optimizations anyway.

timonsku commented:
@6by9 Please see updated commits. checkpatch should be happy except for license header which does not seem to be wanted for overlays.
I enabled the I2C interface in the overlay as well, while nothing technically needs it it seems sensible for a generic overlay to allow access in case there is anything attached that you might want to talk to from userspace.

The only thing unaddressed is the location of the data-lanes property. See my comments on the last review.
If you feel like things should move then I'll revise that.
If not I think things should hopefully have no more form issues. Thanks for bearing with me there.

6by9 commented:
> I enabled the I2C interface in the overlay as well, while nothing technically needs it it seems sensible for a generic overlay to allow access in case there is anything attached that you might want to talk to from userspace.

If you don't need it in the overlay, don't enable it.

i2c0mux will claim GPIOs 0&1 which may be used for other purposes. Your overlay would be incompatible with that for no good reason.
There's already `dtparam=i2c_vc=on` to enable it via a 1-liner, or you can use `dtoverlay=i2c0,pins44_45` if you want just the single I2C port with no mux.

> The only thing unaddressed is the location of the data-lanes property. See my comments on the last review. If you feel like things should move then I'll revise that. If not I think things should hopefully have no more form issues. Thanks for bearing with me there.

data-lanes is a standard property, and it goes in the endpoint as covered by https://www.kernel.org/doc/Documentation/devicetree/bindings/media/video-interfaces.yaml

If you have a bridge or similar DSI device, then you have an input port and an output port, each with their embedded endpoint. The number of lanes is property of the port/endpoint, not the overall device. 
Whilst you only have 1 port, the property remains under the endpoint for consistency.

timonsku commented:
> i2c0mux will claim GPIOs 0&1 which may be used for other purposes. Your overlay would be incompatible with that for no good reason. There's already `dtparam=i2c_vc=on` to enable it via a 1-liner, or you can use `dtoverlay=i2c0,pins44_45` if you want just the single I2C port with no mux.

That is a fair point, will revert.

> If you have a bridge or similar DSI device, then you have an input port and an output port, each with their embedded endpoint. The number of lanes is property of the port/endpoint, not the overall device. Whilst you only have 1 port, the property remains under the endpoint for consistency.

Okay yea, the disconnect to the other bus properties feels a bit odd but fair enough if there is precedence for having lanes defined in the endpoint. Will revise.

timonsku commented:
Pushed the change. I'm a bit uncertain if the bindings are defined correctly this way but given that there should only ever be a single endpoint this feels like the correct approach to me.
I'm relying on the address for port and endpoint to default to 0, afaik they don't need to/shouldn't be defined specifically if the expectation is that there can't be multiple instances?

arturo182 commented on arch/arm/boot/dts/overlays/Makefile:
```diff
@@ -309,7 +309,8 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	wittypi.dtbo \
-	wm8960-soundcard.dtbo
+	wm8960-soundcard.dtbo \
+	panel-dsi-generic.dtbo
```
These seem to be sorted alphabetically and putting this at the end of the list breaks the sort.

timonsku replied:
oh true

6by9 commented on arch/arm/boot/dts/overlays/panel-dsi-generic-overlay.dts:
```diff
@@ -0,0 +1,113 @@
+			};
+			panel: panel-dsi-generic@0 {
+				// See panel-dsi.yaml binding
+				compatible = "panel-dsi";
```
Your bindings state
> Shall contain a panel specific compatible and "panel-dsi" in that order

except you haven't included a panel specific compatible, dummy or otherwise.

timonsku replied:
heh right, yea I'm was sure if it would be sensible to have dummy string there but yea for the sake of consistency it would probably be better to have a dummy string to the generic overlay

6by9 commented on arch/arm/boot/dts/overlays/panel-dsi-generic-overlay.dts:
```diff
@@ -0,0 +1,113 @@
+		rgb565 = <&panel>, "dsi-color-format=RGB565";
+		rgb666p = <&panel>, "dsi-color-format=RGB666_PACKED";
+		rgb666 = <&panel>, "dsi-color-format=RGB666";
+		rgb888 = <&panel>, "dsi-color-format=RGB888";
```
Not needed as this is the default already present in the panel definition.

pelwell replied:
I'm OK with having parameters that select values that are also the defaults. It can take some of the guesswork out of reading a config file.

timonsku replied:
That was sort of my thinking, a lot of people who would use the generic overlay might never see the dts file and just follow a guide. Being explicit in the parameters seems like it would avoid potential confusion.

6by9 commented on arch/arm/boot/dts/overlays/panel-dsi-generic-overlay.dts:
```diff
@@ -0,0 +1,113 @@
+
+	fragment@5 {
+		target = <&panel>;
+		__dormant__ {
```
Why is this dormant? 

As per earlier review comments:
> Seeing as we don't really support command mode (I've never seen a command mode panel in the flesh), you can assume MODE_VIDEO is set. Command mode panels would be supported through [panel-dsi-cm](https://github.com/raspberrypi/linux/blob/rpi-6.1.y/drivers/gpu/drm/panel/panel-dsi-cm.c) anyway, and not panel-simple.
>
> So just set MODE_VIDEO and have done with it.

timonsku replied:
right, I wanted to possibly keep LP video mode as an option and see if it is equivalent to video mode and LP flag, then decided against it and forgot to just remove all of that entirely.
Will prune.

6by9 commented on arch/arm/boot/dts/overlays/Makefile:
```diff
@@ -177,6 +177,7 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	ov5647.dtbo \
 	ov7251.dtbo \
 	ov9281.dtbo \
+	panel-dsi-generic.dtbo \
```
Please follow the pattern of the other dsi and dpi panels and name the overlay vc4-kms-dsi-generic-overlay.dts.

6by9 replied:
Your overlay is now named vc4-kms-dsi-generic, so the Makefile needs to match or it won't get built.

6by9 reviewed (commented):
All overlays must have an entry in the README file documenting the overlay and the associated overrides. In this case most of it can be copied from vc4-kms-dpi-generic.

Icing on the cake would have been to include configuration of a backlight node (as vc4-kms-dpi-generic does), but I'm not going to worry about that one now.

timonsku commented:
I'm with you for the backlight but I'm not sure what the best options for that are. Unlike for dpi you are not guaranteed to have a connection to the 40pin header and you might not wanna touch that with a default if you don't know if there isn't likely some form of HAT on there.
With Pi5 we now also have these two AUX IO but I'm not sure what your plans are for that and if they support PWM?
Happy to do a second PR later on when got hardware that could test that.

Have pushed the requested changes except for leaving RGB888 in as an option. If there is disagreement let me know.

6by9 commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -173,6 +173,30 @@ Params:
+        drm_fb2_rp1_dsi1        Assign /dev/fb2 to the RP1 DSI1 output
+
+        drm_fb2_vc4             Assign /dev/fb2 to the vc4 outputs
+
```
Something funny has gone on with your push. These are part of 61b138adaeaddefe749f421a3b69c67d4a49a8e3, and the mcp23017 changes are part of 27f0e0e195568c06f21ce380f0736bdd219baf3c.

Your branch is based on feb4ea51e77af09c1a4c7c3e5dd33237750f95cd from 5th Oct.

Please do `git fetch` and `git rebase` to pull in all updates that have been made upstream, and that should resolve the mismatch with these erroneous fragments.

6by9 reviewed (commented):
Makefile and weird rebase fixes required.

For vc4-kms-dpi-generic the backlight is disabled by default, so you could have done the same here. If the panel implementation has suitable backlight control then the wiring is sorted via overrides.
(For those designing hardware, an I2C controlled backlight makes most sense - TI look to have a number of options in the LP855x range).
Leave it for now - this has rumbled on long enough.

timonsku commented:
oh yea sorry, divergent feature branch, looks clean to me now. Makefile fixed.

And yea agree, I2C backlight probably the best for now and what we've been using.

6by9 commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -4712,6 +4712,35 @@ Params: clock-frequency         Display clock frequency (Hz)
+                                This is the default
+        two-lane                Use two DSI lanes for data transmission
+        three-lane              Use three DSI lanes for data transmission
+                                Only supported on Pi5+ and CM
```
Pi5+ ?

timonsku replied:
As in for Pi5 and "going forward" assuming that a Pi6 would keep with the 4 channels.

timonsku commented:
Glad to hear and thanks for the guidance!
Should I remove the comment for 3 and 4 channel option before you pull it in?

6by9 commented:
dtoverlaycheck ensures:
- README is in alphabetical order. vc4-kms-dsi-generic doesn't come between vc4-kms-dpi-generic and vc4-kms-dpi-hyperpixel2r
- All parameters are documented, and all documented parameters exist. The overlay uses XXX_lane whilst the documentation says XXX-lane. Hyphen vs underscore. Hyphens are preferred.

If you could update the overlay to use hyphens, correct the alphabetical order, and drop the erroneous `+` from `5B+`, then I think we're there.

timonsku commented:
oh yea, thats what I get for typing it out..
dsi0 was indeed also not in the readme. Pushed fix.

pelwell commented:
Yay - thanks for persevering and getting this over the line.
